packages/backend/src/tools/changelog/diff.test.ts

```diff
@@ -0,0 +1,634 @@
+            },
+          ],
+        },
+      },
+    }
+
+    const previousOutput = {
+      chain: 'ethereum',
+      blockNumber: 1000,
+      contracts: [previousEndpoint, previousUlnV2],
+    } as unknown as DiscoveryOutput
+
+    const currentOutput = {
+      chain: 'ethereum',
+      blockNumber: 2000,
+      contracts: [currentEndpoint, currentUlnV2],
+    } as unknown as DiscoveryOutput
+
+    const changelogEntries = getDiscoveryChanges(previousOutput, currentOutput)
+
+    expect(changelogEntries.added).toEqual([])
+    expect(changelogEntries.removed).toEqual([])
+    expect(changelogEntries.modified).toEqual([
+      {
+        targetName: endpoint.name,
+        targetAddress: endpoint.address,
+        chainId: ChainId.ETHEREUM,
+        blockNumber: currentOutput.blockNumber,
+        operation: 'MODIFY_CONTRACT',
+        type: 'OBJECT_EDITED_PROPERTY',
+        parameterName: 'BLOCK_VERSION',
+        parameterPath: ['BLOCK_VERSION'],
+        previousValue: '1',
+        currentValue: '2',
+      },
+      {
+        targetName: endpoint.name,
+        targetAddress: endpoint.address,
+        chainId: ChainId.ETHEREUM,
+        blockNumber: currentOutput.blockNumber,
+        operation: 'MODIFY_CONTRACT',
+        type: 'OBJECT_EDITED_PROPERTY',
+        parameterName: 'defaultSendVersion',
+        parameterPath: ['defaultSendVersion'],
+        previousValue: '2',
+        currentValue: '3',
+      },
+      {
+        targetName: endpoint.name,
+        targetAddress: endpoint.address,
+        chainId: ChainId.ETHEREUM,
+        blockNumber: currentOutput.blockNumber,
+        operation: 'MODIFY_CONTRACT',
+        type: 'OBJECT_NEW_PROPERTY',
+        parameterName: 'isReceivingPayload',
+        parameterPath: ['isReceivingPayload'],
+        previousValue: null,
+        currentValue: 'false',
+      },
+      {
+        targetName: ulnV2.name,
+        targetAddress: ulnV2.address,
+        chainId: ChainId.ETHEREUM,
+        blockNumber: currentOutput.blockNumber,
+        operation: 'MODIFY_CONTRACT',
+        type: 'OBJECT_EDITED_PROPERTY',
+        parameterName: 'CONFIG_TYPE_INBOUND_BLOCK_CONFIRMATIONS',
+        parameterPath: ['CONFIG_TYPE_INBOUND_BLOCK_CONFIRMATIONS'],
+        previousValue: '20',
+        currentValue: '15',
+      },
+      {
+        targetName: ulnV2.name,
+        targetAddress: ulnV2.address,
+        chainId: ChainId.ETHEREUM,
+        blockNumber: currentOutput.blockNumber,
+        operation: 'MODIFY_CONTRACT',
+        type: 'ARRAY_NEW_ELEMENT',
+        parameterName: 'defaultAdapterParams',
+        parameterPath: ['defaultAdapterParams', '101', '1'],
+        previousValue: null,
+        currentValue:
+          '{"proofType":2,"adapterParams":"0x00010000000000000000000000000000000000000000000000000000000000030d48"}',
+      },
+    ])
+  })
+
+  it('handles contract creation', () => {
+    const currentEndpoint = {
```

Comment:
Just a suggestion: a case with the empty contract (no values) could be added for docs purposes